docs: local testing guide (#829)
nearai/nearai: aws_runner/LOCAL_TESTING.md

# NEAR AI Local Testing Guide

This guide explains how to set up and run the NEAR AI Runner locally. Running the runner locally provides full emulation of the production environment, allowing for proper debugging and fixing of errors that might not be fully reported in the production runner by design.

## Why Use the Local Runner?

By using the local runner, you can:

- See all errors from the agent, API, and UI.

- Run agents directly from your local disk without needing to update versions in the NEAR AI registry.

- Add necessary packages to local runner frameworks ([contact us](http://t.me/nearaialpha) if you need these packages added to the production runners later).

## Prerequisites

Before starting, ensure you have the following installed:

- Docker
- jq (for processing JSON in shell commands)
- A configured NEAR AI account

## Setting Up the Local Runner

1. Clone the repository and navigate to the root directory:
   ```sh
   git clone https://github.com/nearai/nearai.git
   cd nearai
   ```

> 

Reviewer: ```suggestion

2. Install `hub` dependencies

```sh
pip install -e ".[hub]"
````

```

Reviewer: I ran into a lot of missing dependencies when trying to running `fastapi dev app.py --port 8081` later in this guide

> 2. Build the Docker image for the desired framework:

Reviewer: ```suggestion
3. Build the Docker image for the desired framework:
```

   - **Base framework:**
     ```sh
     docker build -f aws_runner/Dockerfile --platform linux/amd64 --build-arg FRAMEWORK=-base -t nearai-runner:test .
     ```

> 3. Start the Docker container and mount all your local agents to agent registry:

Reviewer: ```suggestion
4. Start the Docker container and mount all your local agents to agent registry:
```

   ```sh
   docker run --platform linux/amd64 -p 9009:8080 -v ~/.nearai/registry:/root/.nearai/registry nearai-runner:test
   ```
   This will start the runner on port `9009`.

    Note: Sometimes, you may need to restart the local Docker instance to refresh files mounted from the local agents registry.

> 4. Run a local SingleStore Database:

Reviewer: ```suggestion
5. Run a local SingleStore Database:
```

```bash
docker run -d --name singlestoredb-dev \
           -e ROOT_PASSWORD="change-me" \
           --platform linux/amd64 \
           -p 3306:3306 -p 8080:8080 -p 9000:9000 \
           ghcr.io/singlestore-labs/singlestoredb-dev:latest
```

Set database environment variables in the `hub/.env` file:

```shell
DATABASE_HOST=localhost
DATABASE_USER=root
DATABASE_PASSWORD=change-me
DATABASE_NAME=hub
```

> 

Reviewer: I would add this step here:

Connect to SingleStore database using the following command:

```bash
# Connect to the running container
docker exec -it singlestoredb-dev mysql -u root -p'choose-a-password'
```

Create the hub database:

```bash
CREATE DATABASE hub;
exit;
```

> Apply all migrations using alembic:

Reviewer: ```suggestion
6. Now apply all migrations using alembic in the `/hub` directory:
```

```bash

> # Install alembic
> pip install alembic
> # Apply migrations
> alembic upgrade head

Reviewer: ```suggestion
# Navigate to the hub directory
cd hub
# Install alembic
pip install alembic
# Apply migrations
alembic upgrade head
```

```

## Running the Local Hub

> To use the local runner with the NEAR AI Hub, update the `hub/.env` file:

Reviewer: ```suggestion
To use the local runner with the NEAR AI Hub, create a `hub/.env` file or rename and use `hub/.env.example`
```

```
RUNNER_ENVIRONMENT="custom_runner"
CUSTOM_RUNNER_URL=http://localhost:9009/2015-03-31/functions/function/invocations
API_URL=http://host.docker.internal:8081
```
Then, start the local NEAR AI Hub API:
```sh
cd hub
fastapi dev app.py --port 8081
```

> 

Reviewer: ```suggestion

(TIP: If you received a `_lzma` error this commonly happens on macOS. To fix, try `brew install xz` and reinstall Python 3.11. Then run `pip install -e ".[hub]"` from the root of this project)
```

Reviewer: another edge case I ran into. Thought it might be helpful for others.

## Running the Local Hub UI

1. Navigate to the Hub UI directory:
   ```sh
   cd hub/demo
   ```

2. Update the `hub/demo/.env` file to use the local runner:
   ```

>    ROUTER_URL=http://127.0.0.1:8081/v1

Reviewer 2: Shouldn't it be: `host.docker.internal`? iirc on Macs loopback will fail - but `host.docker.internal` is not enabled by default on Linux hosts...

Reviewer 2: Disregard the above, we don't run the UI in Docker...
